### DISCUSSION TITLE
Regarding the Null pointer derefernce in calloc

### DISCUSSION THREAD

**Discussion by anzosasuke** [fkie-cad/cwe_checker#344] [Q&A]
So i was analyzing the binary using this checker, 

1. in 001026aa, it says null pointer dereference, it is a potential dereference, right and not the actual null pointer dereferencing?
2. In fig2, the calloc function shows "potential overflow due to multiplication before call to calloc at 00107187" - CWE190, how is this different from 1? They both triggered different sets of errors, why?, I can see some IMUL happening in calloc at 00107187, but why there was no NULL pointer dereference?

Any insights on this would 
be so helpful. Thank you!

![fig1](https://user-images.githubusercontent.com/33804183/179849915-374465aa-703e-4a53-b393-7afabdc328c2.png)

![fig2](https://user-images.githubusercontent.com/33804183/179850379-27ee0bdb-8ee3-4b6c-835e-458784c84577.png)

**Comment by Enkelmann** [comment]
Regarding 1.: Yes, all warnings of the cwe_checker are only potential bugs, since false positives are almost not avoidable for this type of static analysis. And if I read your code snippet correctly, then your example here is a false positive, since the analysis checks in line 72 in the decompiler window whether `calloc` returned a null pointer here.

Regarding 2.: The `IMUL` instruction triggered the CWE190 warning, which also looks like a false positive. Actually, this might be a case where the current master branch of the cwe_checker does not produce a false positive here, since we now check whether we can statically ensure that only small integers got multiplied. But the analysis should also generate a CWE-476 Null Pointer Dereference warning at the same address. At least if the shown decompiler snippet is correct there is no Null check here. Can you check that with the command line output of the cwe_checker? The corresponding Ghidra EOL comment in your screenshot may have been overwritten by the CWE-190 warning at the same address.

**Reply by anzosasuke** [reply]
Thanks @Enkelmann . I used the current master branch and the CWE190 warning did go away, but i didn't get the Null Pointer warning. I checked both the command line and json output.
I have another example here which should have a Null pointer warning but didn't. I can't seem to grasp the reason why?
If I could know the reason behind it, I would be grateful;. 

[assembly.txt](https://github.com/fkie-cad/cwe_checker/files/9151548/assembly.txt)

![ghidradecompiler_output](https://user-images.githubusercontent.com/33804183/180022569-992ec851-69d5-4e4c-b849-46afd8bd9ac7.png)

![source1](https://user-images.githubusercontent.com/33804183/180023642-222d2268-15aa-4ccf-b332-3326618b6e40.png)

**Reply by Enkelmann** [reply]
Well, the CWE-476 check only looks for unchecked return values of function calls that may return null pointers. Your example does not contain such a function call, thus it cannot find this type of Null dereference.

If I read your example correctly, the Null pointer comes from a parameter to the function itself that may be a Null pointer, right? In theory we could improve the CWE-476 check to also find these kinds of Null dereferences. That would depend on the Memory check being precise enough to tell us when a pointer parameter could be a Null pointer. I would have to thoroughly test that, because if it is not yet precise enough for that we would only generate lots of false positives.

**Reply by Enkelmann** [reply]
Why the previous example is not identified as a possible Null dereference is unclear to me. I think I would have to debug the binary in question to find the cause for that. Could very well be a bug in the CWE-476 check.

**Reply by anzosasuke** [reply]
Thanks @Enkelmann. Here is the binary if you'd like to find the bug
[mcf_s.zip](https://github.com/fkie-cad/cwe_checker/files/9162420/mcf_s.zip)

function name:primal_net_simplex

**Reply by anzosasuke** [reply]
Here I have another example which is different from above, I think.
This is the source code
![weird2](https://user-images.githubusercontent.com/33804183/180311047-14809092-2c35-40e2-b9ce-2e4cd1e227fc.png)
Cwe_checker do not find this as well. Why is it so?(I am definitely not trying to criticize. It's for my Research.). As always, I am grateful for your insights.
here is in ghidra.
![weirdghidra](https://user-images.githubusercontent.com/33804183/180311717-2ce6040f-7b21-4b8e-8aa1-519fb8356c12.png)

**Reply by Enkelmann** [reply]
Regarding the `mcf_s` binary: I found the reason for the check failing to find the Null dereference: `basket` is a global variable. Back when I wrote the check it was still very unclear how we wanted to track global variables. Nowadays we have a plan for that, but it is not yet implemented. So back then there was an alternative (and quite hacky) way to track global variables just for this check. But if I remember correctly, it produced too many false positives, so it was disabled again. I will put this on my TODO list for when the check gets reworked (probably as soon as general tracking of global variables is implemented).

Regarding your last example: The CWE-476 check cannot find this kind of Null dereference, since the Null pointer is not a return value of some function call. The Memory check could in theory find this kind of Null dereference, but it tries to be very conservative before flagging something as Null dereference. At the moment it usually flags only cases where some pointer is explicitly set to Null. And since it already produces too many false positives the general analysis quality has to be improved before we try to find more Null dereferences with it.

Long story short: In both cases we need to improve analysis quality before we can track the bug variant without producing too many false positive warnings in the process.
